pkg/providers/cloudstack/validator.go

```diff
@@ -48,83 +55,122 @@ type ProviderCmkClient interface {
+	azNamesToCheck := []string{}
+	if len(datacenterConfig.Spec.Domain) > 0 {
+		azNamesToCheck = append(azNamesToCheck, decoder.CloudStackGlobalAZ)
+	}
+	for _, az := range datacenterConfig.Spec.AvailabilityZones {
+		azNamesToCheck = append(azNamesToCheck, az.Name)
 	}
-	logger.MarkPass("Connected to server")
 
+	for _, azName := range azNamesToCheck {
+		cmk := v.cmks[azName]
+		if cmk == nil {
+			return fmt.Errorf("cannot find CloudStack profile for availability zone %s", azName)
+		}
+		if err := cmk.ValidateCloudStackConnection(ctx); err != nil {
+			return fmt.Errorf("failed validating connection to cloudstack %s: %v", azName, err)
+		}
+	}
+
+	logger.MarkPass("Connected to", "servers", azNamesToCheck)
 	return nil
 }
 
 func (v *Validator) ValidateCloudStackDatacenterConfig(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if len(datacenterConfig.Spec.Domain) <= 0 {
-		return fmt.Errorf("CloudStackDatacenterConfig domain is not set or is empty")
-	}
-	if datacenterConfig.Spec.ManagementApiEndpoint == "" {
-		return fmt.Errorf("CloudStackDatacenterConfig managementApiEndpoint is not set or is empty")
-	}
-	_, err := getHostnameFromUrl(datacenterConfig.Spec.ManagementApiEndpoint)
-	if err != nil {
-		return fmt.Errorf("checking management api endpoint: %v", err)
-	}
-	execConfig, err := decoder.ParseCloudStackSecret()
-	if err != nil {
-		return fmt.Errorf("parsing cloudstack secret: %v", err)
-	}
-	if execConfig.ManagementUrl != datacenterConfig.Spec.ManagementApiEndpoint {
-		return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
-			execConfig.ManagementUrl, datacenterConfig.Spec.ManagementApiEndpoint)
-	}
-
-	if err := v.validateDomainAndAccount(ctx, datacenterConfig); err != nil {
+	if err := v.generateLocalAvailabilityZones(ctx, datacenterConfig); err != nil {
 		return err
 	}
 
-	zones, errZone := v.cmk.ValidateZonesPresent(ctx, datacenterConfig.Spec.Zones)
-	if errZone != nil {
-		return fmt.Errorf("checking zones %v", errZone)
-	}
+	for _, az := range v.availabilityZones {
+		_, err := getHostnameFromUrl(az.ManagementApiEndpoint)
+		if err != nil {
+			return fmt.Errorf("checking management api endpoint: %v", err)
+		}
 
-	for _, zone := range datacenterConfig.Spec.Zones {
-		if len(zone.Network.Id) == 0 && len(zone.Network.Name) == 0 {
-			return fmt.Errorf("zone network is not set or is empty")
+		cmk := v.cmks[az.Name]
+		if cmk.GetManagementApiEndpoint(ctx) != az.ManagementApiEndpoint {
+			return fmt.Errorf("cloudstack secret management url (%s) differs from cluster spec management url (%s)",
+				cmk.GetManagementApiEndpoint(ctx), az.ManagementApiEndpoint)
 		}
-		err := v.cmk.ValidateNetworkPresent(ctx, domainId, zone, zones, datacenterConfig.Spec.Account, len(zones) > 1)
+
+		zones, err := cmk.ValidateZonesPresent(ctx, []anywherev1.CloudStackZone{az.CloudStackAvailabilityZone.Zone})
```

Comment:
it seems like we should refactor this to only check if a single zone is present, not multiple. That'll require less error handling below as well